src/fsharp/FSharp.Core/seq.fs

```diff
@@ -1377,29 +1377,21 @@ namespace Microsoft.FSharp.Collections
                       invalidArg "source" (SR.GetString(SR.notEnoughElements))
                   while e.MoveNext() do
                       yield e.Current }
-
+                           
         [<CompiledName("Last")>]
         let last (source : seq<_>) =
             checkNonNull "source" source
-            use e = source.GetEnumerator()
-            if e.MoveNext() then
-                let mutable res = e.Current
-                while (e.MoveNext()) do res <- e.Current
-                res
-            else
-                invalidArg "source" LanguagePrimitives.ErrorStrings.InputSequenceEmptyString
-
+            match Microsoft.FSharp.Primitives.Basics.Seq.tryLastV source with
+            | ValueSome x -> x
+            | ValueNone -> invalidArg "source" LanguagePrimitives.ErrorStrings.InputSequenceEmptyString
+        
         [<CompiledName("TryLast")>]
         let tryLast (source : seq<_>) =
             checkNonNull "source" source
-            use e = source.GetEnumerator()
-            if e.MoveNext() then
-                let mutable res = e.Current
-                while (e.MoveNext()) do res <- e.Current
-                Some res
-            else
-                None
-
+            match Microsoft.FSharp.Primitives.Basics.Seq.tryLastV source with
```

Comment:
that's hilarious pattern match ;-)

@dsyme @cartermp  can we have this as a helper function or something?

Reply:
I am open to opening namespace.

Reply:
Yep, let's open em

Reply:
Not talking about the namespace. I meant a "cast" from ValieOption to Option.

Reply:
I think that’s how Don intended it. See above

Reply:
Its also to ensure that the same exception is raised as for the other types

Reply:
I don't see any reason not to have `Option.toValueOption` and `ValueOption.toOption`. I guess open a suggestion and RFC on it